### PR TITLE
Recover jobs when broker restarted

### PR DESCRIPTION
**Pull request by jmrodri**
* update app to kick off recover before starting to listen to requests
* refactored extractedCredentials to store the podname so we can recover them
* handle in progress state
* fixed the make build & deploy workflow
* made homedir in Dockerfile writable so oc login would work in the container

### CODE REVIEW

**Comment by jmrodri** [comment]
TEST
----

* Deploy broker to your cluster
```
git checkout restart
oc cluster up
catasb/archived/stage2/create_cluster_user.sh zeus
make build PROJECT=yourdockerorg
docker push docker.io/yourdockerorg/ansible-service-broker-asb:latest
make deploy
```
* Prepare another terminal to delete the broker pod (DO NOT RUN IT YET)
```
oc delete pods asb-NAMEOFPOD
```
* Connect to broker, and start a provision, maybe the `postgresql-demo-apb`
```
./asbcli connect asb-1338-ansible-service-broker.127.0.0.1.nip.io
# t
# p
```
Choose something to provision.
* RUN THE `oc delete` FROM OTHER TERMINAL BEFORE PROVISION FINISHES.
* Now look at the new asb pod created.
```
oc logs asb-NEWPODNAME
```
* Look for `Initiating Recovery Process`

```
[2017-05-31T00:41:19.807Z] [NOTICE] Initiating Recovery Process
[2017-05-31T00:41:19.807Z] [INFO] Listening for provision messages
[2017-05-31T00:41:19.807Z] [NOTICE] Entered Recover
[2017-05-31T00:41:19.808Z] [DEBUG] Dao::FindByState
[2017-05-31T00:41:19.811Z] [DEBUG] Successfully loaded [ 1 ] jobstate objects from etcd dir [ /state/ ]
[2017-05-31T00:41:19.812Z] [DEBUG] Found! jobstate [{b534703c-ca3d-4af4-af90-b64843a36ee2 in progress aa-fa5b72d1-d22d-
4e6b-b553-b75bcf9178a5}] matched given state: [in progress].
[2017-05-31T00:41:19.812Z] [NOTICE] ---------------- Refresh login
[2017-05-31T00:41:19.813Z] [DEBUG] Refreshing login token...
[2017-05-31T00:41:19.813Z] [DEBUG] target: [ 172.17.0.1:8443 ]
[2017-05-31T00:41:19.813Z] [DEBUG] user: [ zeus ]
[2017-05-31T00:41:19.813Z] [DEBUG] password:[ zeus ]
[2017-05-31T00:41:20.077Z] [DEBUG] No error reported after running oc login. Cmd output:
[2017-05-31T00:41:20.077Z] [DEBUG] Login successful.
You have access to the following projects and can switch between them with 'oc project <projectname>':
    ansible-service-broker
  * default
    kube-system
    myproject
    openshift
    openshift-infra
Using project "default".
Welcome! See 'oc help' to get started.

[2017-05-31T00:41:20.077Z] [NOTICE] We have a pod: aa-fa5b72d1-d22d-4e6b-b553-b75bcf9178a5
[2017-05-31T00:41:20.077Z] [DEBUG] Calling monitorOutput on aa-fa5b72d1-d22d-4e6b-b553-b75bcf9178a5
[2017-05-31T00:41:23.174Z] [DEBUG] oc log output: Login successful.
You have access to the following projects and can switch between them with 'oc project <projectname>':
    ansible-service-broker
  * default
    kube-system
    myproject
    openshift
    openshift-infra
Using project "default".
Welcome! See 'oc help' to get started.
[2017-05-31T00:41:20.077Z] [NOTICE] We have a pod: aa-fa5b72d1-d22d-4e6b-b553-b75bcf9178a5
[2017-05-31T00:41:20.077Z] [DEBUG] Calling monitorOutput on aa-fa5b72d1-d22d-4e6b-b553-b75bcf9178a5
[2017-05-31T00:41:23.174Z] [DEBUG] oc log output: Login successful.
You have access to the following projects and can switch between them with 'oc project <projectname>':
    ansible-service-broker
  * default
    kube-system
    myproject
    openshift
    openshift-infra
Using project "default".
Welcome! See 'oc help' to get started.
PLAY [Deploy curlcreds-apb to openshift] ***************************************
TASK [setup] *******************************************************************
ok: [localhost]
TASK [curlcreds-apb-openshift : encode credentials] ****************************
changed: [localhost]
 [WARNING]: Consider using get_url or uri module rather than running curl
TASK [curlcreds-apb-openshift : debug] *****************************************
ok: [localhost] => {
    "msg": "<BIND_CREDENTIALS></BIND_CREDENTIALS>"
}
PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0   

[2017-05-31T00:41:23.174Z] [DEBUG] broker::Recover, got ExtractedCredentials!
[2017-05-31T00:41:23.174Z] [DEBUG] Dao::SetRaw [ /state/14a306ff-000e-4fa9-b266-f22e7f7017e9/job/b534703c-ca3d-4af4-af9
0-b64843a36ee2 ] -> [ {"token":"b534703c-ca3d-4af4-af90-b64843a36ee2","state":"succeeded","podname":"aa-fa5b72d1-d22d-4
e6b-b553-b75bcf9178a5"} ]
[2017-05-31T00:41:23.175Z] [DEBUG] Dao::SetRaw [ /extracted_credentials/14a306ff-000e-4fa9-b266-f22e7f7017e9 ] -> [ {} 
]
[2017-05-31T00:41:23.177Z] [NOTICE] Leaving Recover
[2017-05-31T00:41:23.177Z] [NOTICE] recover called
[2017-05-31T00:41:23.177Z] [NOTICE] Ansible Service Broker Started
[2017-05-31T00:41:23.177Z] [NOTICE] Listening on http://0.0.0.0:1338
```

**Comment by jmrodri** [comment]
We should merge https://github.com/fusor/ansible-service-broker/pull/124 *BEFORE* merging this PR. I'll have to rebase afterwards.

**Comment by jmrodri** [comment]
I will rebase this early next week.

**Comment by jmrodri** [comment]
REBASED!

**Comment by rthallisey** [comment]
[(13) Handle restarts while persisting state [asb][sp5] [13]](https://trello.com/c/FArNWgS0/191-13-handle-restarts-while-persisting-state-asbsp5-13)

File: pkg/apb/ext_creds.go [hunk 1]
```diff
@@ -75,7 +75,7 @@ func extractCredentials(
 }
 
 // HACK: this really is a crappy way of getting output
-func getPodName(output []byte, log *logging.Logger) (string, error) {
+func GetPodName(output []byte, log *logging.Logger) (string, error) {
```
Comment (rthallisey) [inline pkg/apb/ext_creds.go]: This isn't used because we get the podname  back when we create the apb

File: pkg/apb/ext_creds_test.go [hunk 2]
```diff
@@ -210,7 +210,7 @@ func assertError(t *testing.T, err error, verifystr string) {
 func TestGetPodName(t *testing.T) {
 	output := []byte(`pod "aa-03be586b-25fc-4d68-8591-51deed8765a1" created`)
 
-	podname, _ := getPodName(output, nil)
+	podname, _ := GetPodName(output, nil)
```
Comment (rthallisey) [inline pkg/apb/ext_creds_test.go]: remove

File: pkg/app/app.go [hunk 3]
```diff
@@ -131,7 +131,27 @@ func CreateApp() App {
 	return app
 }
 
+// TODO: should this be a go routine?
```
Comment (rthallisey) [inline pkg/app/app.go]: Recovery is hard enough. Let's master that before we move to something like concurrent recovery.

File: pkg/app/app.go [hunk 4]
```diff
@@ -131,7 +131,27 @@ func CreateApp() App {
...
+	a.log.Notice(msg)
+}
+
 func (a *App) Start() {
+	a.log.Info("Initiating Recovery Process")
+	// TODO: probably return an error or some sort of message such that we can
+	// see if we need to go any further. Do I even need a method on the App?
+	// Should this just be called from main before we call Start? I feel like
+	// main shouldn't give a crap if we need to recover or not.
+	a.Recover()
```
Comment (rthallisey) [inline pkg/app/app.go]: One thing that's not ideal is we run recovery every time without choice.  I'm worried that if a recovery breaks on something a user will never be able to get past here.

File: pkg/dao/dao.go [hunk 5]
```diff
@@ -186,6 +188,56 @@ func (d *Dao) BatchGetSpecs(dir string) ([]*apb.Spec, error) {
 	return specs, nil
 }
 
+// FindByState - Retriieve all the jobs that match state
```
Comment (rthallisey) [inline pkg/dao/dao.go]: misspelling

File: pkg/broker/broker.go [hunk 6]
```diff
@@ -116,6 +118,129 @@ func (a AnsibleBroker) Bootstrap() (*BootstrapResponse, error) {
...
+		// since we're here we have an in progress job
+		instance, err := a.dao.GetServiceInstance(rs.InstanceId.String())
+
+		// do we have a podname?
+		if rs.State.Podname == "" {
+			// NO, we do not have a podname
+
+			a.log.Info(fmt.Sprintf("No podname. Attempting to restart job: %s", rs.InstanceId.String()))
+
+			si, err := a.dao.GetServiceInstance(rs.InstanceId.String())
```
Comment (shawn-hurley) [inline pkg/broker/broker.go]: Can you just use the instance already retrieved? or should we wait to retrieve to here? I don't think we need both.

Comment (rthallisey) [inline pkg/broker/broker.go]: Ya we can use the instance from 173

File: pkg/broker/broker.go [hunk 7]
```diff
@@ -116,6 +118,129 @@ func (a AnsibleBroker) Bootstrap() (*BootstrapResponse, error) {
...
+					* extractCreds if available
+				no
+					* create a monitoring job to update status
+	*/
+
+	// let's see if we need to recover any of these
+	for _, rs := range recoverStatuses {
+
+		// since we're here we have an in progress job
+		instance, err := a.dao.GetServiceInstance(rs.InstanceId.String())
```
Comment (rthallisey) [inline pkg/broker/broker.go]: ```InstanceId.String()``` is used over and over again. Let's make it a var.

File: pkg/broker/broker.go [hunk 8]
```diff
@@ -116,6 +118,129 @@ func (a AnsibleBroker) Bootstrap() (*BootstrapResponse, error) {
...
+
```
Comment (rthallisey) [inline pkg/broker/broker.go]: Also not catching an error.

File: etc/ex.dev.config.yaml [hunk 9]
```diff
@@ -20,5 +20,5 @@ openshift:
 broker:
   dev_broker: true
   launch_apb_on_bind: false
+  recovery: true
```
Comment (dymurray) [inline etc/ex.dev.config.yaml]: Can you please add this to the template too? templates/deploy-ansible-service-broker.yaml under configMap. https://github.com/openshift/ansible-service-broker/blob/master/templates/deploy-ansible-service-broker.template.yaml#L211

Comment (eriknelson) [inline etc/ex.dev.config.yaml]: Also need it in run_template.sh I think.

File: pkg/broker/broker.go [hunk 10]
```diff
@@ -117,6 +120,128 @@ func (a AnsibleBroker) Bootstrap() (*BootstrapResponse, error) {
...
+	a.log.Debug("Recovery refresh login")
+
+	// need a client to refresh the login
+	client, cerr := apb.NewClient(a.log)
+	if cerr != nil {
+		a.log.Error("recovery failed: could not create client during recovery")
+		return "", cerr
+	}
+
+	cerr = client.RefreshLoginToken(a.clusterConfig)
```
Comment (shawn-hurley) [inline pkg/broker/broker.go]: We need to check if this is in cluster or not.

Comment (rthallisey) [inline pkg/broker/broker.go]: I don't think we need to make this call at all. The token gets refreshed when we attempt to make any contact with the cluster. Deleting this.

**Comment by eriknelson** [comment]
ACK!

Started postgres provision, deleted asb pod, 2nd broker pod came up and recovered. I then provisioned hello-world and was able to bind to the postgres instance.

**Pull request merged**